pyogrio/_io.pyx

```diff
@@ -2418,6 +2528,29 @@ def ogr_write(
 
         log.info(f"Created {num_records:,} records" )
 
+        # close dataset to force driver to flush data
+        exc = ogr_close(ogr_dataset)
+        ogr_dataset = NULL
+        if exc:
+            raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+        ### Copy in-memory file back to path_or_fp object
+        if vsi_handle != NULL:
+            # rewind to beginning of file and read bytes back from memory
+            VSIFSeekL(vsi_handle, 0, 0)
+
+            # Take ownership of the buffer to avoid a copy; GDAL will automatically
+            # unlink the memory file
```

Comment:
Should we then set `vsi_handle` to null to avoid we unlink it afterwards again? (although that doesn't seem to be giving errors right now)